library(dplyr)
library(readr)

#this script is to plot the loci with candidate genes

################################# settings
#you need to adjust this before running

#day number for easier identification of outputs
daynumber = "20240618"
#do not use slashes or points, since this will be used for the output folder name

#daynumber of the file generated by 7_LD_chromosome
daynumber_LD = "20240607"
#daynumber of 9_gene_analysis output
daynumber_genes = "20240530"

#pool settings:
species = "wheat"
#barley or wheat
poolName = "YR_pooled"
#barley pools:
#NFNB_pooled, PM_2015, Blr_pooled, Scald_pooled, SFNB_pooled
#wheat pools:
#YR_pooled, LR_2017, SR_2017, YLS_2017, STB_2019, YS_2019

locus_name = "w4a1b"
#use existing locus name
#loci:
#Blr:"b6h1"
#NFNB:"b7h2","b6h3","b5h1"
#PM:"b1h1","b6h2"
#Scald (MAF 5%):"b3h1","b7h1"
#YR:"w3b1","w4a1","w4a1a","w4a1b"
#YS:"w2a1"

range_by_genes = TRUE #if true, gene list will determine range, otherwise the option below will
#if the above is set to false, choose:
SNP_amount = 100 #how many SNPs around the most significant SNP should be included

################################# prepare

#import data
locus_info = read.csv(paste0("results/",species,"/gene_analysis/",daynumber_genes,"_",poolName,"/locus_info.csv"))
sig_SNP_loci = read.csv(paste0("results/",species,"/gene_analysis/",daynumber_genes,"_",poolName,"/sig_SNP_loci.csv"))

locus_info = filter(locus_info, locus == locus_name)
sig_SNP_loci = filter(sig_SNP_loci, locus == locus_name)

chromosome = sig_SNP_loci[1,2]
nei_dist = sig_SNP_loci[1,5] #neighbour distance of most significant SNP
focal_SNP_pos = sig_SNP_loci[1,4] #position of most significant SNP

#import data
LD_s = read.csv(paste0("results/",species,"/LD/chromosome/",daynumber_LD,"_",poolName,"/SNPwise_LD_of_chromosome_",chromosome,".csv"))
neiGWAS_results = read.csv(paste0("results/",species,"/neiGWAS/data_collection/neiGWAS_results_",poolName,".csv"))
selfGWAS_results = read.csv(paste0("results/",species,"/selfGWAS_compare/data_collection/selfGWAS_results_",poolName,".csv"))
transcript_list = read.csv(paste0("results/",species,"/gene_analysis/",daynumber_genes,"_",poolName,"/",locus_name,"_transcript_list.csv"))

rownames(LD_s)<- colnames(LD_s)
pvalues_n = neiGWAS_results[,c("chr","pos",paste0("p_log10_",nei_dist))]
pvalues_s = selfGWAS_results[,c("chr","pos","p_log10_self")]
pvalues_n = arrange(pvalues_n, pos)
pvalues_n = arrange(pvalues_n, chr)
pvalues_s = arrange(pvalues_s, pos)
pvalues_s = arrange(pvalues_s, chr)
transcripts = transcript_list[,c(2,3,6,7)]
transcripts = mutate(transcripts, "name"=paste0(gene_ID,".",transcript_ID))
transcripts = transcripts[,c(5,3,4)]
colnames(transcripts)<-c("name","left","right")

#get SNP range and cut transcripts
pvalues_c = mutate(pvalues_s, ID = 1:nrow(pvalues_s))
pvalues_c = pvalues_c[pvalues_c$chr==chromosome,]
chr_beg = pvalues_c$ID[1]-1
if(range_by_genes){
  genes_min = min(transcripts$left)
  genes_max = max(transcripts$right)
  SNPs_left = pvalues_c[pvalues_c$pos<genes_min,]
  if(nrow(SNPs_left)>0){
    SNP_min = SNPs_left[nrow(SNPs_left),"ID"] #this should be SNP ID, not pos
  }else{
    SNP_min = pvalues_c$"ID"[1]
  }
  SNPs_right = pvalues_c[pvalues_c$pos>genes_max,]
  if(nrow(SNPs_right)>0){
    SNP_max = SNPs_right[1,"ID"]
  }else{
    SNP_max = pvalues_c$"ID"[ncol(pvalues_c)]
  }
  rm(SNPs_left)
  rm(SNPs_right)
}else{
  pvalues_c = mutate(pvalues_c, dist = abs(pos-focal_SNP_pos))
  pvalues_c = arrange(pvalues_c, dist)
  if(nrow(pvalues_c)>SNP_amount){
    pvalues_c = pvalues_c[1:SNP_amount,]
  }
  SNP_min = min(pvalues_c$ID)
  SNP_max = max(pvalues_c$ID)
  transcripts = transcripts[which(transcripts$right>pvalues_n$pos[SNP_min]&transcripts$left<pvalues_n$pos[SNP_max]),]
}
v_LD = LD_s[(SNP_min-chr_beg):(SNP_max-chr_beg),(SNP_min-chr_beg):(SNP_max-chr_beg)]
v_logp = pvalues_n[SNP_min:SNP_max,3]
v_logp_self = pvalues_s[SNP_min:SNP_max,3]
SNP_amount = length(v_logp)

#clean
rm(LD_s)
rm(neiGWAS_results)
rm(selfGWAS_results)
rm(pvalues_c)

#prepare for plot
LD_part_max = SNP_max-SNP_min+1 #amount of SNPs in range
LD_part_min = 0
plot_title = paste0("candidate-gene-mapping-plot of locus ",locus_name)
logp_scaled = v_logp/max(v_logp)
SNP_names = rownames(v_LD)
SNP_pos = pvalues_n[SNP_min:SNP_max,2]/1000
transcripts$left = transcripts$left/1000
transcripts$right = transcripts$right/1000
transcripts_offset = runif(n = nrow(transcripts), min = 0.0, max = 1.0)
maxlogp = round(max(v_logp),digits=3)
maxlogp_dist = SNP_pos[which(v_logp==max(v_logp))]
#get significance levels
sig_levels = c(-log10(0.05/nrow(pvalues_n)), -log10(0.01/nrow(pvalues_n)), -log10(0.001/nrow(pvalues_n)))
sig_names = c("level 1: 0.05 b.c. 5%","level 2: 0.01 b.c.","level 3: 0.001 b.c.")
SNP_sigtype = {}
for(i in 1:length(v_logp)){
  if(v_logp[i]<sig_levels[1]){
    SNP_sigtype = c(SNP_sigtype, 0)
  }
  else if(v_logp[i]<sig_levels[2]){
    SNP_sigtype = c(SNP_sigtype, 1)
  }
  else if(v_logp[i]<sig_levels[3]){
    SNP_sigtype = c(SNP_sigtype, 2)
  }
  else{
    SNP_sigtype = c(SNP_sigtype, 3)
  }
}
#get borders of each SNP area
inter_SNP_pos = SNP_pos[1] + ((SNP_pos[1] - SNP_pos[2])/2)
for(i in 1:(length(SNP_pos)-1)){
  inter_SNP_pos = c(inter_SNP_pos, (SNP_pos[i] + SNP_pos[i+1])/2)
}
inter_SNP_pos = c(inter_SNP_pos, SNP_pos[length(SNP_pos)] + ((SNP_pos[length(SNP_pos)] - SNP_pos[length(SNP_pos)-1])/2))
#get range
pos_min = inter_SNP_pos[1] #kb
pos_max = inter_SNP_pos[length(inter_SNP_pos)] #kb
symbol_radius_x = (pos_max-pos_min)/320 #160
symbol_radius_y = (LD_part_max/160)
symbol_offset = (-LD_part_max/50)
#get sig SNPs
sig_SNPs = which(SNP_sigtype!=0)

if(!file.exists(paste0("results/",species,"/gene_mapping"))){
  dir.create(path = paste0("results/",species,"/gene_mapping"))
}

################################# candidate gene mapping plot

png(filename = paste0("results/",species,"/gene_mapping/",locus_name,"_gene_mapping_plot.png"),width=3600,height=3000,units="px")

bgc <- par(bg = "white")
par(mfrow = c(3,1))
layout(matrix(c(1,1,1,1,1,2,2,2,3),nrow = 9,ncol = 1))
par(mar = c(2.0,4.3,8.1,9.1))
par(col.main = "black",col.axis = "black", col.lab = "black")
plot(c(pos_min, pos_max), c(LD_part_min, LD_part_max), type = "n", xlab = "", ylab = "LD (red) and -log(p) (green)", main = plot_title, cex.main = 4, yaxt = "n") + axis(side = 4, at = (c(1:length(SNP_names))-0.5),labels = SNP_names, las = 2) + axis(side = 1, col = "white")
for(j in 1:length(SNP_names)){
  for(i in 1:length(SNP_names)){
    rect(inter_SNP_pos[i], j-0.8, inter_SNP_pos[i+1], j-0.2, col = rgb(1.0, 1.0 - (v_LD[i,j]), 1.0 - (v_LD[i,j])), lwd = 0)
    rect(inter_SNP_pos[i], j-0.2, inter_SNP_pos[i+1], j-0.0, col = rgb(1.0 - logp_scaled[i], 1.0, 1.0 - logp_scaled[i]), lwd = 0)
  }
  for(k in 1:nrow(transcripts)){
    rect(transcripts$left[k], j-1.0, transcripts$right[k], j-0.8, col = rgb(0.2, 0.2, 1.0, 0.5), lwd = 0)
  }
  if(SNP_sigtype[j]==1){
    poly_x = c(SNP_pos[j]-symbol_radius_x,SNP_pos[j],SNP_pos[j]+symbol_radius_x,SNP_pos[j])
    poly_y = c(symbol_offset,symbol_offset-symbol_radius_y,symbol_offset,symbol_offset+symbol_radius_y)
    polygon(poly_x,poly_y,col="black",border=FALSE)
  }
  else if(SNP_sigtype[j]==2){ # a star
    poly_x = c(SNP_pos[j]+(symbol_radius_x*sinpi(0.0)),SNP_pos[j]+(symbol_radius_x*sinpi(0.2)*0.382),SNP_pos[j]+(symbol_radius_x*sinpi(0.4)),SNP_pos[j]+(symbol_radius_x*sinpi(0.6)*0.382),SNP_pos[j]+(symbol_radius_x*sinpi(0.8)),SNP_pos[j]+(symbol_radius_x*sinpi(1.0)*0.382),SNP_pos[j]+(symbol_radius_x*sinpi(1.2)),SNP_pos[j]+(symbol_radius_x*sinpi(1.4)*0.382),SNP_pos[j]+(symbol_radius_x*sinpi(1.6)),SNP_pos[j]+(symbol_radius_x*sinpi(1.8)*0.382))
    poly_y = c(symbol_offset+(symbol_radius_y*cospi(0.0)),symbol_offset+(symbol_radius_y*cospi(0.2)*0.382),symbol_offset+(symbol_radius_y*cospi(0.4)),symbol_offset+(symbol_radius_y*cospi(0.6)*0.382),symbol_offset+(symbol_radius_y*cospi(0.8)),symbol_offset+(symbol_radius_y*cospi(1.0)*0.382),symbol_offset+(symbol_radius_y*cospi(1.2)),symbol_offset+(symbol_radius_y*cospi(1.4)*0.382),symbol_offset+(symbol_radius_y*cospi(1.6)),symbol_offset+(symbol_radius_y*cospi(1.8)*0.382))
    polygon(poly_x,poly_y,col="black",border=FALSE)
  }
  else{ # a fake circle
    poly_x = c(SNP_pos[j]+(symbol_radius_x*sinpi(0.0)*0.6),SNP_pos[j]+(symbol_radius_x*sinpi(0.2)*0.6),SNP_pos[j]+(symbol_radius_x*sinpi(0.4)*0.6),SNP_pos[j]+(symbol_radius_x*sinpi(0.6)*0.6),SNP_pos[j]+(symbol_radius_x*sinpi(0.8)*0.6),SNP_pos[j]+(symbol_radius_x*sinpi(1.0)*0.6),SNP_pos[j]+(symbol_radius_x*sinpi(1.2)*0.6),SNP_pos[j]+(symbol_radius_x*sinpi(1.4)*0.6),SNP_pos[j]+(symbol_radius_x*sinpi(1.6)*0.6),SNP_pos[j]+(symbol_radius_x*sinpi(1.8)*0.6))
    poly_y = c(symbol_offset+(symbol_radius_y*cospi(0.0)*0.6),symbol_offset+(symbol_radius_y*cospi(0.2)*0.6),symbol_offset+(symbol_radius_y*cospi(0.4)*0.6),symbol_offset+(symbol_radius_y*cospi(0.6)*0.6),symbol_offset+(symbol_radius_y*cospi(0.8)*0.6),symbol_offset+(symbol_radius_y*cospi(1.0)*0.6),symbol_offset+(symbol_radius_y*cospi(1.2)*0.6),symbol_offset+(symbol_radius_y*cospi(1.4)*0.6),symbol_offset+(symbol_radius_y*cospi(1.6)*0.6),symbol_offset+(symbol_radius_y*cospi(1.8))*0.6)
    polygon(poly_x,poly_y,col="black",border=FALSE)
  }
}
for(k in 1:nrow(transcripts)){
  rect(transcripts$left[k], length(SNP_names), transcripts$right[k], length(SNP_names)+0.2, col = rgb(0.2, 0.2, 1.0, 0.5), lwd = 0)
}
mtext(paste0("phenotype: ",poolName,", range: ",chromosome,":",1000*SNP_pos[1],"-",chromosome,":",1000*SNP_pos[length(SNP_pos)],", SNPs: ",length(SNP_pos),", max -log(p): ",maxlogp,", transcripts: ",nrow(transcripts)), col = "black")

par(mar = c(2.0,4.3,0.0,9.1))
plot(c(pos_min, pos_max), c(0.0, maxlogp), type = "n", xlab = "", ylab = "-log(p_nei) (white), scaled LD (red) and -log(p_self) (green)", yaxt = "n") + axis(side = 2, col = "black") + axis(side = 1, col = "black") + axis(side = 4, at = sig_levels,labels = sig_names, las = 2)
for(k in 1:nrow(transcripts)){
  rect(transcripts$left[k], 0.0, transcripts$right[k], maxlogp, col = rgb(0.2, 0.2, 1.0, 0.5), lwd = 0)
}
abline(h = sig_levels[1], col = rgb(0.5,0.5,0.5))
abline(h = sig_levels[2], col = rgb(0.5,0.5,0.5))
abline(h = sig_levels[3], col = rgb(0.5,0.5,0.5))
for(i in sig_SNPs){
  lpp = v_LD[,i]*v_logp[i]
  lines(x = SNP_pos, y = lpp, col = rgb(1.0,0.0,0.0))
}
lines(x = SNP_pos, y = v_logp_self, col = rgb(0.0,1.0,0.0))
lines(x = SNP_pos, y = v_logp, col = "white")

par(mar = c(9.1,4.3,15.0,9.1))
plot(c(pos_min, pos_max), c(-length(sig_SNPs), 4), type = "n", xlab = "", ylab = "transcripts (blue) with ID, SNPs -log(p_nei) (green) and LD with sig. SNPs (red)", xaxt = "n", yaxt = "n") + axis(side = 4, at = (c(1:length(sig_SNPs))-length(sig_SNPs)-0.5),labels = SNP_names[sig_SNPs], las = 2) + axis(side = 3, at = (transcripts$left+transcripts$right)/2, labels = transcripts$name, las = 2, mgp = c(3.0,0.0,0.0))# + axis(side = 1, at = SNP_pos, labels = SNP_names, las = 2, mgp = c(3.0,0.0,0.0))
for(k in 1:nrow(transcripts)){
  rect(transcripts$left[k], 2.0 + transcripts_offset[k], transcripts$right[k], 3.0 + transcripts_offset[k], col = rgb(0.2, 0.2, 1.0, 0.5), border = "black")
}
for(j in 1:length(SNP_names)){
  poly_x = c(SNP_pos[j]-(0.5*symbol_radius_x),SNP_pos[j],SNP_pos[j]+(0.5*symbol_radius_x))
  poly_y = c(0.5,1.5,0.5)
  polygon(poly_x, poly_y, col = rgb(0.0, logp_scaled[j], 0.0), border="black")
  for(i in 1:length(sig_SNPs)){
    poly_y = c(-i,1-i,-i)
    polygon(poly_x, poly_y, col = rgb(1.0, 1.0 -(v_LD[j,sig_SNPs[1+length(sig_SNPs)-i]]), 1.0 - (v_LD[j,sig_SNPs[1+length(sig_SNPs)-i]])), border="black")
  }
}
text(x = SNP_pos, y = par("usr")[3] - 0.25, srt = -90, adj = 0, labels = SNP_names, xpd = TRUE, col = "black")

dev.off()

################################# candidate gene mapping plot 2

png(filename = paste0("results/",species,"/gene_mapping/",locus_name,"_gene_mapping_plot_alt.png"),width=3600,height=2000,units="px")

bgc <- par(bg = "white")
par(mfrow = c(8,1))
layout(matrix(c(1,2,3,3,3,4,5,5,5,6,7,7,7,8,9,9,9),nrow = 17,ncol = 1))
par(mar = c(2.1,4.3,10.1,9.1))
par(col.main = "black",col.axis = "black", col.lab = "black")
plot(c(0.0, 1.0), c(0.0, 1.0), type = "n", xlab = "", ylab = "", main = plot_title, cex.main = 4, yaxt = "n", xaxt = "n")
mtext(paste0("phenotype: ",poolName,", range: ",chromosome,":",1000*SNP_pos[1],"-",chromosome,":",1000*SNP_pos[length(SNP_pos)],", SNPs: ",length(SNP_pos),", max -log(p): ",maxlogp,", transcripts: ",nrow(transcripts)), col = "black")
pos_dif = pos_max - pos_min
for(j in 1:4){
  par(mar = c(2.1,4.3,6.1,9.1))
  plot(c(pos_min + (pos_dif*(j-1)/4), pos_min + (pos_dif*j/4)), c(0.0, 2.0), type = "n", xlab = "", ylab = "candidate transcripts (blue)", yaxt = "n") + axis(side = 1, col = "black", at = c(floor((pos_min + (pos_dif*(j-1)/4))/100):ceiling((pos_min + (pos_dif*j/4))/100))*100, xpd = TRUE) + axis(side = 3, at = (transcripts$left+transcripts$right)/2, labels = transcripts$name, las = 2, mgp = c(3.0,0.0,0.0))
  for(k in 1:nrow(transcripts)){
    rect(transcripts$left[k], transcripts_offset[k], transcripts$right[k], 1.0 + transcripts_offset[k], col = rgb(0.2, 0.2, 1.0, 0.5), border = "black")
  }
  par(mar = c(9.1,4.3,0.1,9.1))
  plot(c(pos_min + (pos_dif*(j-1)/4), pos_min + (pos_dif*j/4)), c(0.0, maxlogp), type = "n", xlab = "", ylab = "-log(p_nei) (black), scaled LD (red) and -log(p_self) (green)", yaxt = "n", xaxt = "n") + axis(side = 2, col = "black") + axis(side = 4, at = sig_levels,labels = sig_names, las = 2)
  for(i in 0:ceiling(1.2*maxlogp)){
    abline(h = i, col = rgb(0.8,0.8,0.8))
  }
  for(k in 1:nrow(transcripts)){
    rect(transcripts$left[k], 0.0, transcripts$right[k], maxlogp, col = rgb(0.2, 0.2, 1.0, 0.5), lwd = 0)
  }
  abline(h = sig_levels[1], col = rgb(0.5,0.5,0.5))
  abline(h = sig_levels[2], col = rgb(0.5,0.5,0.5))
  abline(h = sig_levels[3], col = rgb(0.5,0.5,0.5))
  for(i in sig_SNPs){
    lpp = v_LD[,i]*v_logp[i]
    lines(x = SNP_pos, y = lpp, col = rgb(1.0,0.0,0.0))
  }
  lines(x = SNP_pos, y = v_logp_self, col = rgb(0.0,1.0,0.0))
  lines(x = SNP_pos, y = v_logp, col = "black")
  text(x = SNP_pos, y = par("usr")[3] - 0.25, srt = -90, adj = 0, labels = SNP_names, xpd = TRUE, col = "black")
}

dev.off()

#clean
rm(bgc)
rm(locus_info)
rm(pvalues_n)
rm(pvalues_s)
rm(sig_SNP_loci)
rm(transcript_list)
rm(transcripts)
rm(v_LD)
rm(chr_beg)
rm(chromosome)
rm(daynumber_LD)
rm(daynumber_genes)
rm(focal_SNP_pos)
rm(genes_max)
rm(genes_min)
rm(i)
rm(j)
rm(k)
rm(inter_SNP_pos)
rm(LD_part_max)
rm(LD_part_min)
rm(locus_name)
rm(logp_scaled)
rm(lpp)
rm(maxlogp)
rm(maxlogp_dist)
rm(nei_dist)
rm(plot_title)
rm(poly_x)
rm(poly_y)
rm(pos_dif)
rm(pos_max)
rm(pos_min)
rm(range_by_genes)
rm(sig_levels)
rm(sig_names)
rm(sig_SNPs)
rm(SNP_amount)
rm(SNP_max)
rm(SNP_min)
rm(SNP_names)
rm(SNP_pos)
rm(SNP_sigtype)
rm(symbol_offset)
rm(symbol_radius_x)
rm(symbol_radius_y)
rm(transcripts_offset)
rm(v_logp)
rm(v_logp_self)

